### ISSUE TITLE
QueryResult will lose the column metadata

### ISSUE THREAD

**Issue by gaoqiangz** [prisma/tiberius#165]
sql:
```sql
CREATE PROCEDURE p_test
AS
BEGIN

	---uncomment this line,will be ok
	---SET NOCOUNT ON;

	---or comment this statement,also will be ok
	IF 1 = 0
	BEGIN
		SELECT 'test' AS COL
		RETURN;
	END

	SELECT 'test' AS COL
END
GO
```
rust:
```rust
let rs = client.simple_query("EXEC p_test").await?;
println!("{:?}",rs.columns()); //None
```

**Comment by pimeys** [comment]
Yep. Bug is a bug, fixing :)

I'm actually also thinking of rewriting the whole stream handling very soon... The way it's currently done is quite hacky and we should follow rust guidelines a bit better with it. Stay tuned.

**Comment by gaoqiangz** [comment]
:+1:

**Comment by pimeys** [comment]
And, meanwhile, you should be able to use methods such as `into_results` to get the correct columns. I'm working on a new API today, but this is going to be a breaking change for stream handling and will be a major version change.

**Comment by pimeys** [comment]
Hey @gaoqiangz, check out my PR above and see how do you like the interface. Your example changes a bit, but this gives the correct result:

```rust
#[tokio::main]
async fn main() -> anyhow::Result<()> {
    use futures::TryStreamExt;

    let config = Config::from_ado_string(&CONN_STR)?;

    let tcp = TcpStream::connect(config.get_addr()).await?;
    tcp.set_nodelay(true)?;

    let mut client = Client::connect(config, tcp.compat_write()).await?;
    let mut rs = client.simple_query("EXEC p_test").await?;

    while let Some(item) = rs.try_next().await? {
        match item {
            QueryItem::Metadata(meta) => {
                println!("Metadata: {:?}", meta);
            }
            QueryItem::Row(row) => {
                println!("Row: {:?}", row);
            }
        }
    }

    Ok(())
}
```

**Comment by gaoqiangz** [comment]
It looks good, I have tested it without problems.

**Comment by gaoqiangz** [comment]
In most cases, the query contains only one resultset, which may provide a simpler interface for these most cases.
In my project case,I will get the meta info to check out if the result has the expected columns before fetch the rest rows.

**Comment by pimeys** [comment]
There is a nice edge case for the new architecture, that allows me to actually implement `fn metadata()` to the stream... That might help in your case. I'll write it next, seeing how it feels like...

**Comment by pimeys** [comment]
If you pull my latest changes from the branch, I added two new things to the `QueryResult`:

- `columns` method
- `into_row_stream` method that converts the stream into a `Stream<Item = Row>`, skipping the metadata for non-MARS use-cases.

**Comment by pimeys** [comment]
Released in v0.6.0 https://crates.io/crates/tiberius
